### ISSUE TITLE
Add utilities to Augment image datasets

### ISSUE THREAD

**Issue by leonardoaraujosantos** [leonardoaraujosantos/DLMatFramework#20]
## Introduction
Add utilities to do batch augmentation during training, also add functions to calculate the mean pixel/image

### Example on torch/lua
```lua
function ImagenetDataset:preprocess()
   if self.split == 'train' then
      return t.Compose{
         t.RandomSizedCrop(224),
         t.ColorJitter({
            brightness = 0.4,
            contrast = 0.4,
            saturation = 0.4,
         }),
         t.Lighting(0.1, pca.eigval, pca.eigvec),
         t.ColorNormalize(meanstd),
         t.HorizontalFlip(0.5),
      }
   elseif self.split == 'val' then
      local Crop = self.opt.tenCrop and t.TenCrop or t.CenterCrop
      return t.Compose{
         t.Scale(256),
         t.ColorNormalize(meanstd),
         Crop(224),
      }
   else
      error('invalid split: ' .. self.split)
   end
end
``` 

## References
* [Old repo code TransformBatch.m](https://github.com/leonardoaraujosantos/Matlab_CS231N/blob/master/utils/TransformBatch.m)
* [Old repo code getMeanImageOnBatch.m](https://github.com/leonardoaraujosantos/Matlab_CS231N/blob/master/utils/getMeanImageOnBatch.m)
* [Old repo code getPixelMeanOnBatch.m](https://github.com/leonardoaraujosantos/Matlab_CS231N/blob/master/utils/getPixelMeanOnBatch.m)
* [Slides](http://pt.slideshare.net/xavigiro/deep-learning-for-computer-vision-data-augmentation-upc-2016?ref=&smtNoRedir=1)
* [Resnet with Tensorflow](https://github.com/tensorflow/models/tree/master/resnet)
* [Torch PCA Augmentation](https://github.com/Atcold/torch-Machine-learning-with-Torch/tree/master/PCA)
* [Commonly used data augmentation techniques for torch7](https://gist.github.com/farrajota/09b3ef287e6eca981b35c07b4357464e)
* [Blog Galaxy Zoo](http://benanne.github.io/2014/04/05/galaxy-zoo.html)
* [Facebook Transforms.lua](https://github.com/facebook/fb.resnet.torch/blob/e8fb31378fd8dc188836cf1a7c62b609eb4fd50a/datasets/transforms.lua)
* [Facebook Transforms usage on imagenet](https://github.com/facebook/fb.resnet.torch/blob/9b7fe7a12817b7fd562ca2caa49f9080b3bff18d/datasets/imagenet.lua#L80)
* [CNN Tips and tricks](http://lamda.nju.edu.cn/weixs/project/CNNTricks/CNNTricks.html)
* [Compose functions in matlab](https://uk.mathworks.com/company/newsletters/articles/tips-and-tricks-combining-functions-using-anonymous-functions.html)

**Comment by leonardoaraujosantos** [comment]
## Adding some examples of augmentation
[SomeAugmentationExamples.zip](https://github.com/leonardoaraujosantos/DLMatFramework/files/771081/SomeAugmentationExamples.zip)

**Comment by leonardoaraujosantos** [comment]
## Random Crop
Basically we drop some pixels from the image to create some sort of random displacement of the input image, as the semantic of the image will not change (If the crop is not too aggressive) it will force the network to learn different ways to represent the same information, so it will improve the validation accuracy.

## References
* https://uk.mathworks.com/help/images/ref/imcrop.html
* http://stackoverflow.com/questions/15589517/how-to-crop-an-image-in-opencv-using-python
* http://code.activestate.com/lists/python-list/706954/
* http://stackoverflow.com/questions/34574714/random-cropping-data-augmentation-convolutional-neural-networks
* https://www.tensorflow.org/tutorials/deep_cnn
* http://stackoverflow.com/questions/35412197/how-can-i-crop-several-and-random-batches-from-an-image

**Comment by leonardoaraujosantos** [comment]
Seems fine, as it improved 4% validation accuracy on cifar 10
